Source/Widgets/Geocoder/GeocoderViewModel.js

```diff
@@ -225,81 +399,128 @@ define([
         });
     }
 
-    function geocode(viewModel) {
-        var query = viewModel.searchText;
+    function chainPromise(promise, geocoderService, query) {
+        return promise
+            .then(function(result) {
+                if (defined(result) && result.state === 'fulfilled' && result.value.length > 0){
+                    return result;
+                }
+                var nextPromise = geocoderService.geocode(query)
+                    .then(function (result) {
+                        return {state: 'fulfilled', value: result};
+                    });
+                if (defined(nextPromise.otherwise)) {
+                    nextPromise.otherwise(function (err) {
+                        return {state: 'rejected', reason: err};
+                    });
+                } else if (defined(nextPromise.catch)) {
```

Comment:
I assume you are doing this to try and handle multiple promise implementations, but I think we should just simplify the code and always assume the promise was created with our version of when.  Otherwise bad things can happen anyway, for example Bluebird promises have a `cancel` function, which would end up making your code think that `cancel` has been called on successful geocodes because you glob your own `cancel` onto the promise as a boolean. (hopefully that make sense).

We we upgrade to bluebird, these problems should go away, but we should just assume everyone is using Cesium's own `when` library for the time being.